src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.EndpointConstraints;
+using Microsoft.AspNetCore.Routing.Matchers;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    internal class MvcEndpointDataSource : EndpointDataSource
+    {
+        private readonly IActionDescriptorCollectionProvider _actions;
+        private readonly MvcEndpointInvokerFactory _invokerFactory;
+        private readonly IActionDescriptorChangeProvider[] _actionDescriptorChangeProviders;
+        private readonly List<Endpoint> _endpoints;
+
+        private IChangeToken _changeToken;
+
+        public MvcEndpointDataSource(
+            IActionDescriptorCollectionProvider actions,
+            MvcEndpointInvokerFactory invokerFactory,
+            IEnumerable<IActionDescriptorChangeProvider> actionDescriptorChangeProviders)
+        {
+            if (actions == null)
+            {
+                throw new ArgumentNullException(nameof(actions));
+            }
+
+            if (invokerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(invokerFactory));
+            }
+
+            if (actionDescriptorChangeProviders == null)
+            {
+                throw new ArgumentNullException(nameof(actionDescriptorChangeProviders));
+            }
+
+            _actions = actions;
+            _invokerFactory = invokerFactory;
+            _actionDescriptorChangeProviders = actionDescriptorChangeProviders.ToArray();
+
+            _endpoints = new List<Endpoint>();
+
+            InitializeEndpoints();
+        }
+
+        private void InitializeEndpoints()
+        {
+            // note: this code has haxxx. This will only work in some constrained scenarios
+            foreach (var action in _actions.ActionDescriptors.Items)
+            {
+                if (action.AttributeRouteInfo == null)
+                {
+                    // Action does not have an attribute route
+                    continue;
+                }
+
+                RequestDelegate invokerDelegate = (context) =>
+                {
+                    var values = context.Features.Get<IEndpointFeature>().Values;
+                    var routeData = new RouteData();
+                    foreach (var kvp in values)
+                    {
+                        routeData.Values.Add(kvp.Key, kvp.Value);
+                    }
+
+                    var actionContext = new ActionContext(context, routeData, action);
+
+                    var invoker = _invokerFactory.CreateInvoker(actionContext);
```

Comment:
We can avoid closing on `_invokerFactory` as it can be retrieved from `context.RequestServices`. Any specific reason you left it this way? since we cannot test this func anyway, I would not be worried about using ServiceLocator pattern here.

Reply:
Invoker factory really has to be global though. I don't see much of an advantage to making this more complex